docs/plugins/npe2_manifest_specification.md

```diff
@@ -0,0 +1,604 @@
+      title: Save points layer to csv
+  writers:
+    - command: napari.write_points
+      filename_extensions: [".csv"]
+      layer_types: ["points"]
+```
+
+**Example**
+
+Multi-layer writer
+
+```yaml
+name: napari-svg
+contributions:
+  commands:
+    - id: napari-svg.svg_writer
+      python_name: napari-svg.hook_implementations:writer
+      title: Save layers as SVG
+  writers:
+    - command: napari_svg.svg_writer
+      layer_types: ["image*", "labels*", "points*", "shapes*", "vectors*"]
+      filename_extensions: [".svg"]
+```
+
+### Layer type constraints
+
+Given a set of layers, compatible writer plugins are selected based their
+_layer type constraints_.
+
+A writer plugin can declare that it will write between _m_ and _n_ layers of a
+specific type where _0≤m≤n_.
+
+For example:
+
+```
+    image      Write exactly 1 image layer.
+    image?     Write 0 or 1 image layers.
+    image+     Write 1 or more image layers.
+    image*     Write 0 or more image layers.
+    image{k}   Write exactly k image layers.
+    image{m,n} Write between m and n layers (inclusive range). Must have m<=n.
+```
+
+When a type is not present in the list of constraints, that corresponds to a
+writer that is not compatible with that type. For example, a writer declaring:
+
+```
+    layer_types=["image+", "points*"]
+```
+
+would not be selected when trying to write an `image` and a `vector` layer
+because the above only works for cases with 0 `vector` layers.
+
+Note that just because a writer declares compatibility with a layer type does
+not mean it actually writes that type. In the example above, the writer might
+accept a set of layers containing `image`s and `point`s, but the write command
+might just ignore the `point` layers. The writer must return `None` for
+unwritten layers.
+
+### Calling convention
+
+Currently, two calling conventions are supported for writers: single-layer and
+multi-layer writers. When at most one layer can be matched by a writer, it
+must use the single-layer convention. Otherwise, the multi-layer convention
+must be used.
+
+#### multi-layer writer
+
+```python
+def writer_function(
+    path: str, layer_data: List[LayerData]
+) -> List[str]:
+    ...
+```
+
+**Parameters**
+
+path(str)
+: Path to file, directory, or resource (like a url).
+
+layer_data(list of LayerData)
+: Each `LayerData` element is a tuple of `(data,meta,layer_type)`.
+
+**Return**
+
+Returns a list of paths that were successfully written.
+
+#### single-layer writer
+
+```python
+def writer_function(
+    path: str, data, meta
+) -> Optional[str]:
+    ...
+```
+
+**Parameters**
+
+path(str)
+: Path to file, directory, or resource (like a url).
+
+data(array or list of array) : Layer data. Can be N dimensional. If an image
+and meta[‘rgb’] is True then the data should be interpreted as RGB or RGBA. If
+meta[‘multiscale’] is True, then the data should be interpreted as a
+multiscale image.
+
+meta(dict)
+: Layer metadata.
```

Comment:
Just a warning about this name. `Layer.metadata` is an existing public property that refers to a different `dict` than this one, which comes from `Layer.as_layer_data_tuple` (which happens to include `Layer.metadata`).

At a minimum in the docs, it would be good to add a general note/warning about this name conflict. Ideally we would resolve the conflict by renaming one of these two (I'm not sure `meta` and `metadata` is enough of a restriction), but this isn't the PR for that.

Reply:
Would this make it clearer?
```suggestion
: Layer metadata specified as an unstructured `dict`
```

Reply:
Not really, because `Layer.metadata` is also an unstructured dict: https://github.com/napari/napari/blob/fcde7328ac6a561998c1cacaca9dc76492c81b08/napari/layers/base/base.py#L73

Reply:
I'm not sure what to suggest here and it's not something we should spend too much time on right now, so unless someone else feels strongly here, we can just ignore this for now.

Reply:
Right; the problem was the choice of the parameter name "meta".  So when we describe it, we should call it "layer attributes"

Reply:
changed description to 'layer attributes'